File: code/ch06/javascript/transport_inversion.js
// Chapter 6, §6.5.3 (A Development View: Module Hierarchies) — dependency inversion.
// The application layer owns the deliver() shape; WebSocketTransport and a test
// fake both conform to it from below.
// Run: node transport_inversion.js
const assert = require("node:assert");

class MessageRouter {                  // application code sees only the deliver() shape
  constructor(transport) { this.transport = transport; }
  route(message) { this.transport.deliver(message.to, message.body); }
}

class WebSocketTransport {             // infrastructure, conforming from below.
  constructor(socket) { this.socket = socket; }      // No Transport type to declare —
  deliver(to, body) { this.socket.send(`${to}:${body}`); } // any deliver()-shaped
}                                                          // object will do.

class FakeTransport extends Array {    // a two-line test double
  deliver(to, body) { this.push([to, body]); }
}

const fake = new FakeTransport();
new MessageRouter(fake).route({ to: "dana", body: "you are on call" });
assert.deepStrictEqual([...fake], [["dana", "you are on call"]]);

// The real implementation conforms too — fed a stand-in socket here so the
// file runs without a network.
class RecordingSocket extends Array {
  send(data) { this.push(data); }
}
const sock = new RecordingSocket();
new MessageRouter(new WebSocketTransport(sock)).route({ to: "dana", body: "hi" });
assert.deepStrictEqual([...sock], ["dana:hi"]);

console.log("fake and websocket transports both satisfy MessageRouter: OK");

File: code/ch06/python/transport_inversion.py
# Chapter 6, §6.5.3 (A Development View: Module Hierarchies) — dependency inversion.
# The application layer owns the Transport interface; WebSocketTransport and a test
# fake both conform to it from below.
# Run: python3 transport_inversion.py
from typing import Protocol


class Transport(Protocol):                 # owned by the application layer
  def deliver(self, to: str, body: str) -> None: ...


class MessageRouter:                       # application code sees only the interface
  def __init__(self, transport: Transport):
    self._transport = transport

  def route(self, message: dict) -> None:
    self._transport.deliver(message["to"], message["body"])


class WebSocketTransport:                  # infrastructure, conforming from below
  def __init__(self, socket):
    self._socket = socket

  def deliver(self, to: str, body: str) -> None:
    self._socket.send(f"{to}:{body}".encode())


class FakeTransport(list):                 # a two-line test double
  def deliver(self, to, body): self.append((to, body))


fake = FakeTransport()
MessageRouter(fake).route({"to": "dana", "body": "you are on call"})
assert fake == [("dana", "you are on call")]


# The real implementation conforms too — fed a stand-in socket here so the
# file runs without a network.
class RecordingSocket(list):
  def send(self, data): self.append(data)


sock = RecordingSocket()
MessageRouter(WebSocketTransport(sock)).route({"to": "dana", "body": "hi"})
assert sock == [b"dana:hi"]

print("fake and websocket transports both satisfy MessageRouter: OK")
